File: app/packages/app_media_capture_bridge/test/media_capture_generated_wire_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';

void main() {
  final packageRoot = Directory.current.absolute;
  final appRoot = packageRoot.parent.parent;
  final repositoryRoot = appRoot.parent;
  final generatedFile = File.fromUri(
    packageRoot.uri.resolve('lib/src/media_capture_wire.g.dart'),
  );
  final codecFile = File.fromUri(
    packageRoot.uri.resolve('lib/src/media_capture_wire_codec.dart'),
  );

  test('Dart generated Wire output has no drift', () async {
    final result = await Process.run(_dartExecutable(), <String>[
      '--packages=${appRoot.uri.resolve('.dart_tool/package_config.json').toFilePath()}',
      appRoot.uri.resolve('tool/generate_media_capture_wire.dart').toFilePath(),
      '--runtime',
      'dart',
      '--check',
    ], workingDirectory: repositoryRoot.path);

    expect(result.exitCode, 0, reason: '${result.stdout}\n${result.stderr}');
  });

  test('codec consumes every generated payload descriptor', () {
    final generated = generatedFile.readAsStringSync();
    final codec = codecFile.readAsStringSync();
    final payloadIds = RegExp(
      r"_GeneratedWirePayloadDescriptor\(\s*'([^']+)'",
      multiLine: true,
    ).allMatches(generated).map((match) => match.group(1)!).toSet();

    expect(payloadIds, isNotEmpty);
    for (final payloadId in payloadIds) {
      expect(codec, contains("'$payloadId'"), reason: payloadId);
    }
    expect(codec, contains('_generatedMatchesWireFieldPrimitive'));
    expect(codec, contains('_generatedErrorDescriptors'));
    expect(codec, contains('_generatedEnvelopeRequiredKeys'));
  });

  test('generated descriptor coverage matches the shared neutral golden', () {
    final generated = generatedFile.readAsStringSync();
    final golden =
        jsonDecode(
              File.fromUri(
                packageRoot.uri.resolve(
                  'test/contracts/media-capture-v4-v3.golden.json',
                ),
              ).readAsStringSync(),
            )
            as Map<String, Object?>;
    final generation = golden['generation']! as Map<String, Object?>;

    expect(
      RegExp(
        r'Source digest \(SHA-256\): ([0-9a-f]{64})',
      ).firstMatch(generated)!.group(1),
      generation['normalizedDescriptorDigest'],
    );
    expect(generation['wireVersion'], 3);
    expect(
      RegExp(
        r'^const String _generatedMediaCaptureWireMethod',
        multiLine: true,
      ).allMatches(generated).length,
      generation['methodCount'],
    );
    expect(
      RegExp(
        r'^const String _generatedMediaCaptureWireEvent',
        multiLine: true,
      ).allMatches(generated).length,
      generation['eventCount'],
    );
    expect(
      RegExp(
        r'^const String _generatedMediaCaptureWireResult',
        multiLine: true,
      ).allMatches(generated).length,
      generation['resultTypeCount'],
    );
    expect(
      RegExp(
        r'^const String _generatedMediaCaptureWireFailure',
        multiLine: true,
      ).allMatches(generated).length,
      generation['failureTypeCount'],
    );
    expect(
      RegExp(
        r'^const String _generatedMediaCaptureWireError',
        multiLine: true,
      ).allMatches(generated).length,
      generation['errorCount'],
    );
    expect(
      RegExp(
        r'^      _GeneratedWirePayloadDescriptor\(',
        multiLine: true,
      ).allMatches(generated).length,
      generation['payloadDescriptorCount'],
    );
    expect(
      RegExp(
        r'^      _GeneratedWireFieldDescriptor\(',
        multiLine: true,
      ).allMatches(generated).length,
      generation['fieldDescriptorCount'],
    );
  });

  test('public package surface does not expose generated internals', () {
    final barrel = File.fromUri(
      packageRoot.uri.resolve('lib/app_media_capture_bridge.dart'),
    ).readAsStringSync();
    final constants = File.fromUri(
      packageRoot.uri.resolve('lib/src/media_capture_constants.dart'),
    ).readAsStringSync();

    expect(barrel, isNot(contains('media_capture_wire.g.dart')));
    expect(barrel, isNot(contains('media_capture_wire_codec.dart')));
    expect(constants, contains("export 'media_capture_wire_codec.dart'"));
    expect(constants, contains('show'));
    expect(constants, isNot(contains('_Generated')));
  });
}

String _dartExecutable() {
  final flutterRoot = Platform.environment['FLUTTER_ROOT'];
  if (flutterRoot != null && flutterRoot.isNotEmpty) {
    final executable = File.fromUri(
      Directory(flutterRoot).uri.resolve('bin/cache/dart-sdk/bin/dart'),
    );
    if (executable.existsSync()) {
      return executable.path;
    }
  }

  final resolved = File(Platform.resolvedExecutable).absolute;
  if (resolved.uri.pathSegments.last == 'dart') {
    return resolved.path;
  }
  var directory = resolved.parent;
  while (directory.parent.path != directory.path) {
    final executable = File.fromUri(directory.uri.resolve('dart-sdk/bin/dart'));
    if (executable.existsSync()) {
      return executable.path;
    }
    directory = directory.parent;
  }
  return 'dart';
}

File: app/packages/app_media_capture_bridge/android/src/test/kotlin/com/example/media_capture/MediaCaptureGeneratedWireTest.kt
package com.example.media_capture

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class MediaCaptureGeneratedWireTest {
    @Test
    fun generatedIdentifiersDriveTheRuntimeSurface() {
        assertEquals(17, GeneratedMediaCaptureWireMethod.entries.size)
        assertEquals(5, GeneratedMediaCaptureWireEvent.entries.size)
        assertEquals(
            GeneratedMediaCaptureWireMethod.entries.mapTo(linkedSetOf()) { it.wireValue },
            MediaCaptureWireCodec.methods,
        )
        assertEquals(
            GeneratedMediaCaptureWireError.entries.mapTo(linkedSetOf()) { it.wireValue },
            generatedErrorDescriptors.mapTo(linkedSetOf()) { it.code },
        )
    }

    @Test
    fun codecConsumesEveryGeneratedPayloadDescriptor() {
        val source =
            File("src/main/kotlin/com/example/media_capture/MediaCaptureWireCodec.kt")
                .readText()

        generatedPayloadDescriptors.forEach { descriptor ->
            assertTrue(source.contains("\"${descriptor.id}\""), descriptor.id)
        }
        assertTrue(source.contains("generatedMatchesWireFieldPrimitive"))
        assertTrue(source.contains("generatedEnvelopeRequiredKeys"))
        assertTrue(source.contains("generatedErrorDescriptors"))
    }

    @Test
    fun generatedSourceHasStableProvenanceAndNoHostMetadata() {
        val source =
            File("src/main/kotlin/com/example/media_capture/MediaCaptureWire.g.kt")
                .readText()

        assertTrue(source.startsWith("// GENERATED CODE - DO NOT MODIFY BY HAND."))
        assertTrue(source.contains("// Generator: media_capture_wire/1"))
        assertTrue(source.contains("// Source digest (SHA-256):"))
        val userHome = System.getProperty("user.home")
        assertTrue(userHome.isNullOrEmpty() || !source.contains(userHome))
        assertFalse(source.contains("Generated at"))
    }

    @Test
    fun generatedDescriptorCoverageMatchesTheSharedNeutralGolden() {
        val source = File("src/main/kotlin/com/example/media_capture/MediaCaptureWire.g.kt").readText()
        val golden = repositoryFile(
            "app/packages/app_media_capture_bridge/test/contracts/media-capture-v4-v3.golden.json",
        ).readText()
        val digest = Regex("Source digest \\(SHA-256\\): ([0-9a-f]{64})").find(source)?.groupValues?.get(1)

        assertTrue(golden.contains("\"normalizedDescriptorDigest\": \"$digest\""))
        assertTrue(golden.contains("\"methodCount\": ${GeneratedMediaCaptureWireMethod.entries.size}"))
        assertTrue(golden.contains("\"eventCount\": ${GeneratedMediaCaptureWireEvent.entries.size}"))
        assertTrue(golden.contains("\"resultTypeCount\": ${GeneratedMediaCaptureWireResult.entries.size}"))
        assertTrue(golden.contains("\"failureTypeCount\": ${GeneratedMediaCaptureWireFailure.entries.size}"))
        assertTrue(golden.contains("\"errorCount\": ${GeneratedMediaCaptureWireError.entries.size}"))
        assertTrue(golden.contains("\"payloadDescriptorCount\": ${generatedPayloadDescriptors.size}"))
        assertTrue(golden.contains("\"fieldDescriptorCount\": ${generatedMediaCaptureWireFields.size}"))
    }

    private fun repositoryFile(relativePath: String): File {
        var directory = File("").absoluteFile
        while (true) {
            val candidate = File(directory, relativePath)
            if (candidate.isFile) return candidate
            directory = directory.parentFile ?: error("Repository root not found")
        }
    }
}
